### PR TITLE
cpuallocator: add CPU priority levels

### PR DESCRIPTION
**Pull request by marquiz**
Implement three levels of CPU priority (high, normal, low) instead of
binary hi/lo flagging previously used. The aim is to have a simple but a
bit more versatile scheme for evaluating CPU priorities. This scheme
could be further elaborated in the future although it might be hard to
come up with one number that could objectively represent the "priority"
of a CPU in a general case.

The patch tries to preserve the existing behavior closely. There is one
notable difference though: after this patch the existing policies (that
utilize cpuallocator) favor "normal" priority cores over "low" priority
cores when allocating cpus for non-prioritized workloads. This can have
a practical effect, especially if the sockets (CPU packages) in a system
are configured non-uniformly, e.g. so that one socket only has "normal"
cores whereas some others have only "high" and "low" priority cores. In
this case the cpuallocator would try to pack all non-prioritized
workloads on the core(s) that have CPUs with matching priority.
Currently none of the policies utilize the possibility to allocate "low"
priority cpus.

### CODE REVIEW

**Review by klihub — COMMENTED** [review]
Looking pretty good. I only had a few nitpicks and comments.

And one final comment unrelated to this PR but related to cpuallocator...
It would also need to be improved with additional tightest-fit allocation rules
beyond the current topology socket/core/thread hierarchy to get to a more
realistic socket/die/NUMA node/core/thread hierachy:
  - try allocating a full die if the number of requested cpus matches exactly
  - try allocating a full NUMA node if the number of cpus matches exactly
  - only then give up and try allocating mere full cores or threads, and also with these
    - try taking sub-NUMA node number of cores/threads from a single NUMA node,
    - try taking sub-die number of cores/threads from a single die

File: pkg/cpuallocator/allocator.go [hunk 1]
```diff
@@ -446,5 +399,114 @@ func newTopologyCache(sys sysfs.System) topologyCache {
...
+		return "high"
+	case PriorityNormal:
+		return "normal"
+	case PriorityLow:
+		return "low"
+	}
+	return "None"
+}
+
+// cmpCPUSet compares to cpusets in terms of preferred cpu priority. Returns:
```
[the remark this answers is not in the document]
Comment (marquiz) [reply pkg/cpuallocator/allocator.go:490]: Fixed

File: pkg/cpuallocator/allocator.go [hunk 2]
```diff
@@ -446,5 +399,114 @@ func newTopologyCache(sys sysfs.System) topologyCache {
...
+// > 1 cpuset A is preferred
+// < 1 cpuset B is preferred
+// 0 if cpusets are equal in terms of cpu priority
```
[the remark this answers is not in the document]
Comment (marquiz) [reply pkg/cpuallocator/allocator.go:493]: Haha, total brain fart 🧠 💨  Thanks for noting this. Fixed

File: pkg/cpuallocator/allocator.go [hunk 3]
```diff
@@ -446,5 +399,114 @@ func newTopologyCache(sys sysfs.System) topologyCache {
...
+
+			// All cpus NOT in the lowest base frequency bin are considered high prio
+			if bf > freqList[0] {
+				p = PriorityHigh
+			} else {
+				p = PriorityLow
+			}
+		}
+
+		// All cpus NOT in the lowest performance epp are considered high prio
```
[the remark this answers is not in the document]
Comment (marquiz) [reply pkg/cpuallocator/allocator.go:459]: Good point, added a note regarding that. The ordering comes directly from the kernel interface.

File: pkg/cpuallocator/allocator.go [hunk 4]
```diff
@@ -446,5 +399,114 @@ func newTopologyCache(sys sysfs.System) topologyCache {
...
+func (p CPUPriority) String() string {
+	switch p {
+	case PriorityHigh:
+		return "high"
+	case PriorityNormal:
+		return "normal"
+	case PriorityLow:
+		return "low"
+	}
+	return "None"
```
[the remark this answers is not in the document]
Comment (marquiz) [reply pkg/cpuallocator/allocator.go:487]: Thanks! I think it's the latter 😬  Fixed

File: pkg/cpuallocator/allocator.go [hunk 5]
```diff
@@ -446,5 +399,114 @@ func newTopologyCache(sys sysfs.System) topologyCache {
 			c.core[id] = sys.CPU(id).ThreadCPUSet()
 		}
 	}
+
+	c.discoverCpuPriorities(sys)
+
 	return c
 }
+
+func (c *topologyCache) discoverCpuPriorities(sys sysfs.System) {
```
[the remark this answers is not in the document]
Comment (marquiz) [reply pkg/cpuallocator/allocator.go:408]: Ah, thanks. I always forget this nice little "feature" of golang. Fixed

**Comment by marquiz** [comment]
> And one final comment unrelated to this PR but related to cpuallocator...
> It would also need to be improved with additional tightest-fit allocation rules
> beyond the current topology socket/core/thread hierarchy to get to a more
> realistic socket/die/NUMA node/core/thread hierachy:
> 
> * try allocating a full die if the number of requested cpus matches exactly
> * try allocating a full NUMA node if the number of cpus matches exactly
> * only then give up and try allocating mere full cores or threads, and also with these
>   
>   * try taking sub-NUMA node number of cores/threads from a single NUMA node,
>   * try taking sub-die number of cores/threads from a single die

I fully agree on this. Even further, I think it would really make sense at some point to tear this whole thing apart and start from the drawing table. I think the code is currently a bit hard to keep up with and becomes slightly quirkier on every new feature added. Even with this relatively simple PR I managed to iterate over many alternative approaches before settling to this one.

We should also evaluate it from the policy point of view, i.e. how to best serve e.g. the topology-aware policy. There is already #191 about improving the allocator. Maybe that should be changed to re-write. This is actually a subject that would crave for mugs of coffee and a whiteboard.

**Comment by klihub** [comment]
> > And one final comment unrelated to this PR but related to cpuallocator...
> > It would also need to be improved with additional tightest-fit allocation rules
> > beyond the current topology socket/core/thread hierarchy to get to a more
> > realistic socket/die/NUMA node/core/thread hierachy:
> > 
> > * try allocating a full die if the number of requested cpus matches exactly
> > * try allocating a full NUMA node if the number of cpus matches exactly
> > * only then give up and try allocating mere full cores or threads, and also with these
> >   
> >   * try taking sub-NUMA node number of cores/threads from a single NUMA node,
> >   * try taking sub-die number of cores/threads from a single die
> 
> I fully agree on this. Even further, I think it would really make sense at some point to tear this whole thing apart and start from the drawing table. I think the code is currently a bit hard to keep up with and becomes slightly quirkier on every new feature added. Even with this relatively simple PR I managed to iterate over many alternative approaches before settling to this one.
> 
> We should also evaluate it from the policy point of view, i.e. how to best serve e.g. the topology-aware policy. 
>There is already #191 about improving the allocator. Maybe that should be changed to re-write. This is actually a > subject that would crave for mugs of coffee and a whiteboard.

I agree that we'll have to rework this. The resulting allocator should have sane default behavior, it should be possible to override it's built-in CPU picking sorting/scoring/picking logic externally. Otherwise too much policy-specific assumptions creep in (or the API becomes overly complex to use if you try to make everything configurable with flags passed to function calls).

But about the allocator having to primarily serve the topology-aware policy, I tend to disagree. The topology-aware policy is the easiest to please exactly because it is policy-aware. It could get along with a much more stupid allocator with next to no built-in intelligence, for instance one which could look like this (`int`s used instead of `ID`s):
```
type Allocator interface {
    // All of these either fully satisfy the allocations or fail and return the empty set.
    AllocateSockets(count int) cpuset.CPUSet
    AllocateDies(count, socket int) cpuset.CPUSet
    AllocateNUMANodes(count, socket, die int) cpuset.CPUSet
    AllocateCores(count, socket, die, NUMANode int) cpuset.CPUSet
    AllocateThreads(count, socket, die, NUMANode int) cpuset.CPUSet
}
```
Implementing such an allocator is very straightforward because topology-wise everything is pre-picked so the needed built-in intelligence is next to none. Also the topology-aware policy does not really need CPU priority/class awareness from the allocator. Similarly to kernel-isolated isolcpus, it treats CPUs in different clockfreq and/or priority classes (STT* and EPP) as separate subclasses/subpools in each topology tree pool.

It is the other policies which are not inherently topology-aware that need an intelligent allocator. Granted, the topology-aware policy does benefit from an topology-aware allocator, in terms of less/simpler code. Mostly when it cannot do a CPU-wise tightest fitting allocation because the workload asks for a disproportionately large chunk of memory which is not available in any of the tightest fitting topology elements with enough free CPU capacity. But in all the other cases it calls the allocator with a cpuset already confined to the chosen topology pool or an isolated, particular priority or other class-based subpool within that.

**Comment by marquiz** [comment]
> But about the allocator having to primarily serve the topology-aware policy, I tend to disagree.

No, I just wrote my comment hastily. What I tried to say is exactly what you described that the requirements from the policy point of view vary greatly and in terms of topology-aware policy most of the "gymnastics" exercised inside the cpuallocator are unneeded

**Comment by askervin** [comment]
This looks good to me. However, we need to coordinate merging this and PR #598 that uses CPU allocator's old API.

**Review by askervin — APPROVED** [review]
In my perspective this would be good.

(Extra +1 from cmpCPUSet, pretty elegant.)

All CPUs in E2E test VMs seem to be recognized as PriorityNormal - base frequency being 0 as corresponding file is missing at least from ubuntu cloud-VM sysfs.

File: pkg/cpuallocator/allocator.go [hunk 6]
```diff
@@ -199,10 +152,8 @@ func (a *allocatorHelper) takeIdlePackages() {
...
 		func(i, j int) bool {
-			iPref := a.topology.pkg[pkgs[i]].Intersection(a.preferred).Size()
-			jPref := a.topology.pkg[pkgs[j]].Intersection(a.preferred).Size()
-			if iPref != jPref {
-				return iPref > jPref
+			if res := a.topology.cpuPriorities.cmpCPUSet(a.topology.pkg[pkgs[i]], a.topology.pkg[pkgs[j]], a.prefer); res != 0 {
+				return res > 0
 			}
 			return pkgs[i] < pkgs[j]
 		})
```
[the remark this answers is not in the document]
Comment (marquiz) [reply pkg/cpuallocator/allocator.go:159]: Fixed in a separate oneliner patch

**Comment by askervin** [comment]
Commit https://github.com/intel/cri-resource-manager/pull/602/commits/a1e9803cc9b565a232b3cda857a5d3718fa2f712 changes allocation so that quite a few E2E tests break up. For instance,

```
test/e2e$ reinstall_cri_resmgr=1 ./run_tests.sh policies.test-suite/podpools/n4c16/test01-basic-placement
```

I'll investigate how the CPUs are picked up to the pool from the topology, and let's discuss if it's what we want.

**Comment by askervin** [comment]
Topology of "n4c16" e2e test virtual machines:
```
package0 die0 node0 core0 thread0 cpu00
                          thread1 cpu01
                    core1 thread0 cpu02
                          thread1 cpu03
                    mem   node0   2G
              node1 core2 thread0 cpu04
                          thread1 cpu05
                    core3 thread0 cpu06
                          thread1 cpu07
                    mem   node1   2G
package1 die0 node2 core0 thread0 cpu08
                          thread1 cpu09
                    core1 thread0 cpu10
                          thread1 cpu11
                    mem   node2   2G
              node3 core2 thread0 cpu12
                          thread1 cpu13
                    core3 thread0 cpu14
                          thread1 cpu15
                    mem   node3   2G
```
cri-resmgr:
```
...
I0205 09:45:44.849890  310176 policy.go:210] activating 'podpools' policy...
I0205 09:45:44.849984  310176 policy.go:213]   with available resources:
I0205 09:45:44.850070  310176 policy.go:215]     - CPU=#2-15
I0205 09:45:44.850150  310176 policy.go:219]   with reserved resources:
I0205 09:45:44.850202  310176 policy.go:221]     - CPU=#15
I0205 09:45:44.850854  310176 allocator.go:475] discovered 0 priority cpus:
I0205 09:45:44.851079  310176 allocator.go:475] discovered 1 priority cpus: 0-15
I0205 09:45:44.851477  310176 allocator.go:475] discovered 2 priority cpus:
I0205 09:45:44.851532  310176 podpools-policy.go:129] creating podpools policy...
I0205 09:45:44.851572  310176 podpools-policy.go:158] creating podpools configuration
I0205 09:45:44.851648  310176 podpools-policy.go:524] allocating 3 out of 13 non-reserved CPUs for 3 "singlecpu" pools
I0205 09:45:44.851690  310176 allocator.go:141] * takeIdlePackages()...
I0205 09:45:44.851754  310176 allocator.go:161]  => idle packages sorted by preference: []
I0205 09:45:44.851782  310176 allocator.go:182] * takeIdleCores()...
I0205 09:45:44.852033  310176 allocator.go:205]  => idle cores sorted by preference: [2 4 6 8 10 12]
I0205 09:45:44.852086  310176 allocator.go:210]  => considering core 2 (#2-3)...
I0205 09:45:44.852127  310176 allocator.go:210]  => considering core 4 (#4-5)...
I0205 09:45:44.852165  310176 allocator.go:210]  => considering core 6 (#6-7)...
I0205 09:45:44.852202  310176 allocator.go:210]  => considering core 8 (#8-9)...
I0205 09:45:44.852240  310176 allocator.go:210]  => considering core 10 (#10-11)...
I0205 09:45:44.852281  310176 allocator.go:210]  => considering core 12 (#12-13)...
I0205 09:45:44.852434  310176 allocator.go:234]  => idle threads unsorted: [2 3 4 5 6 7 8 9 10 11 12 13 14]
I0205 09:45:44.853689  310176 allocator.go:293]  => idle threads sorted: [14 8 9 10 11 12 13 2 3 4 5 6 7]
I0205 09:45:44.853739  310176 allocator.go:298]  => considering thread 14 (#14-15)...
I0205 09:45:44.853835  310176 allocator.go:363] 1 cpus from #2-14 (preferring #normal) => #14 
...
```

That is, when asked to allocate single CPU, the allocator returns a thread where the other thread (`cpu15`) is allocated to the reserved CPU set. This didn't happen before the commit.

File: pkg/cpuallocator/allocator.go [hunk 7]
```diff
@@ -446,5 +399,115 @@ func newTopologyCache(sys sysfs.System) topologyCache {
...
+	case PriorityLow:
+		return "low"
+	}
+	return "none"
+}
+
+// cmpCPUSet compares two cpusets in terms of preferred cpu priority. Returns:
+//   > 0 if cpuset A is preferred
+//   < 0 cpuset B is preferred
+//   0 if cpusets A and B are equal in terms of cpu priority
```
Comment (klihub) [inline pkg/cpuallocator/allocator.go:494]: typo: < 0 if cpuset B...

Comment (marquiz) [reply pkg/cpuallocator/allocator.go:494]: Fixed

**Comment by marquiz** [comment]
> That is, when asked to allocate single CPU, the allocator returns a thread where the other thread (`cpu15`) is allocated to the reserved CPU set. This didn't happen before the commit.

My reasoning goes like this:
- Previously, the allocator preferred socket 0 over socket 1 because it had fewer available cores
- With this patch cpuallocator started to prefer socket 1 because it had more CPUs of a desired priority level. In this case this was actually a bug, caused by `cmpCPUSet()` that returned non-zero even if `PriorityNone` was specified. I now pushed a new version of the PR that fixes this
- cpuallocator reserved this hyperthread in order to try to save full cpu cores

The problem should now be fixed. Probably the main point/concern is that even really subtle changes can have big effect on how the cpuallocator works. We need that re-write with better unit tests...

**Comment by askervin** [comment]
Thanks @marquiz, just found the same thing in `cmpCPUSet()`. I'll try the new version...

**Comment by askervin** [comment]
The change did not affect selecting cpu14 for the singlecpu[0] pool, because there prefer=PriorityNormal (allocating a user-defined pool).
```
Frame 10: ./cri-resource-manager/pkg/cri/resource-manager/policy/builtin/podpools/podpools-policy.go:529
529:                         cpus, err := p.cpuAllocator.AllocateCpus(freeCpus, cpusPerPool, cpuallocator.PriorityNormal)
```

That is, selection is still made because socket 1 has more (7) CPUs than socket 0 (6) in the intersection with the `from` cpuset, and because of packing tightly reservations of single CPU to cores (`iCoreFree`/`jCoreFree`). As this sounds like a sound reasoning to select the CPUs, I suppose the tests need to be updated.

I'd probably aim to make them a bit more tolerant to CPU allocator detail changes: concentrate on the number of CPUs rather than CPU identities.

Is it ok to you if I start updating the tests in a patch on top of this patch?

**Comment by marquiz** [comment]
> The change did not affect selecting cpu14 for the singlecpu[0] pool, because there prefer=PriorityNormal (allocating a user-defined pool).

Argh, I missed that detail.


> That is, selection is still made because socket 1 has more (7) CPUs than socket 0 (6) in the intersection with the `from` cpuset, and because of packing tightly reservations of single CPU to cores (`iCoreFree`/`jCoreFree`). As this sounds like a sound reasoning to select the CPUs, I suppose the tests need to be updated.
> 
> I'd probably aim to make them a bit more tolerant to CPU allocator detail changes: concentrate on the number of CPUs rather than CPU identities.
> 
> Is it ok to you if I start updating the tests in a patch on top of this patch?

Hold your horses, I think the cpuallocator could be improved. The socket comparison (`cmpCPUSet()`) should take into account the number of cpus requested in addition to the preferred priority. In case of picking single threads, in particular, we should pick the socket that has least available free threads (but enough to satisfy our priority preference), in order to reduce "fragmentation" and preserve larger "chunks" of free cores for bigger workloads.

**Review by klihub — COMMENTED** [review]
Need to run through the full end-to-end test set to see if any of the current test which mostly have their expected CPUSet's hardcoded should be updated.

**Comment by askervin** [comment]
@marquiz , the latest patch makes the allocation sequence look nice in the podpools e2e tests.

I modified podpools e2e tests to be easier to read and maintain when cpuallocator changes - and less fragile in some parts. These tests now pass on your latest commit (https://github.com/intel/cri-resource-manager/pull/602/commits/5a5c9f59488f89bb9a14caea60979713aaec0840):

https://github.com/askervin/cri-resource-manager/commit/6966f60fd398ce91e02d66f36f5d02063c18837a

Do you think it would make sense to include that commit to your patchset?

Didn't check other e2e tests yet...

**Comment by marquiz** [comment]
> Do you think it would make sense to include that commit to your patchset?

I don't have a strong opinion 😄  Maybe you could just submit it as a separate PR

**Review by askervin — APPROVED** [review]
LGTM.
Passes E2E tests with PR #618.

**Review by klihub — APPROVED** [review]

**Pull request merged**